keps/sig-node/2535-ensure-secret-pulled-images/README.md

```diff
@@ -165,57 +201,289 @@ ensure all tenants have rights to the images that are already present on a host.
+MUST be assumed as non-existent. This might cause an addition request to pull an image.
+If it is successful, the original failing record should be overridden to reflect the
+new success.
+
+#### Cache Directory Structure
+
+A new subdirectory is introduced to the kubelet's directory: `image_manager`.
+The directory contains two further subdirectories: `pulling` and `pulled`. The former
+serves for kubelet to store the state about images that it intends to pull. The
+latter subdirectory should contain information about all the images that were
+successfully pulled by the kubelet.
+
+The figure below shows the Go structures that should appear in files in the
+subdirectories described above. These structures **are not part of the REST API**
+but belong to the `imagemanager.kubelet.config.k8s.io` configuration API group.
+They only ever appear as JSON structures in files outside the kubelet memory.
+
+`ImagePullIntent` is the structure that is present in files in the `image_manager/pulling`
+directory. The filename of each such file is a SHA-256 hash of the image spec prepended
+by `sha256~`.
+
+`ImagePulledRecord` is the structure that appears in files in the `image_manager/pulled`
+directory. The filename of each such file is a sanitized image ID provided by the
+container runtime. This KEP expects the image IDs provided to be in the form of `<hash-alg>:<hash-val>`,
+thus the sanitization consists of replacing the `<hash-alg>:` with `<hash-alg>~`.
+
+```go
+type ImagePullIntent struct {
+    metav1.TypeMeta
+
+    ImageSpec string `json:"imageSpec"` // the filename is a hash of this value
+    Credentials ImagePullCredentials `json:"credentials"`
+}
+
+type ImagePulledRecord struct {
+    metav1.TypeMeta
+
+    ImageID string `json:"imageId"` // the filename is a filesystem sanitized image of this value
+    Credentials ImagePullCredentials `json:"credentials"`
+}
+
+// ImagePullCredentials serves two indexes
+// - First index serves for searching for a specific Kubernetes secret that contained
+//   the credentials used to pull an image. This assumes uninterrupted pulls if
+//   credentials were rotated in a Kubernetes secret.
+// - Second index serves for searching for a specific hash of credentials used to pull
+//   an image. This assumes uninterrupted pulls if the same set of credentials is
+//   spread across the cluster.
+type ImagePullCredentials struct {
+    // CredentialHashes is an index of hashes of all the credentials that an image was pulled with
+    // +optional
+    CredentialHashes map[string]ImagePullCredentials `json:"credentialHashes"`
+
+    // KuberneteSecretCoordinates is an index of coordinates of all the kubernetes
+    // secrets that were used to pull the image
+    // +optional
+    KubernetesSecretCoordinates map[KubernetesSecretCoordinates]ImagePullSecret `json:"kubernetesSecretCoordinates"`
+
+    // NodeWideCredentials is a flag denoting that the image was pulled with node-wide
+    // credentials
```

Comment:
if true, is this mutually exclusive with the credentialHashes / kubernetesSecretCoordinates fields?

Reply:
I think so - I'll add that in the comment.

If my understanding is correct, this field would be turned to `true` if the credentials were pulled from the node's [DockerKeyring](https://github.com/kubernetes/kubernetes/blob/19e8e59d06cbdd067a9b5c141b674b63a09d0e5c/pkg/credentialprovider/plugins.go#L50).

[KEP-4412: Projected Service Account Tokens for Kubelet Image Credential Providers](https://github.com/kubernetes/enhancements/pull/4846) suggests the DockerKeyring might change in a way that the credentials would no longer be node-wide for some providers, and so that's something to keep in mind here. I assume a new field would be introduced for those cases.